File: spec/azk/cli/commands/exec_spec.lua
local hh    = require('spec.spec_helper')
local azk   = require('azk')
local app   = require('azk.app')
local shell = require('azk.cli.shell')
local utils = require('azk.utils')
local fs    = require('azk.utils.fs')

local path    = azk.path
local command = require('azk.cli.commands.exec')

describe("Azk #cli #commandexec", function()
  local i18n_f     = azk.i18n.module("command_exec")
  local i18n_app_f = azk.i18n.module("app")
  local i18n_pro_f = azk.i18n.module("provision")

  after_each(function()
    hh.remove_test_images()
  end)

  it("should return 1 with blank invocation", function()
    local result = command.run()
    assert.is.equal(result, 1)
  end)

  it("should require azkfile.json", function()
    local project = utils.tmp_dir()
    fs.cd(project, function()
      local output = shell.capture_io(function()
        command.run("any")
      end)

      local msg = i18n_app_f("no_such", { file = azk.manifest })
      assert.has_log("error", msg, output.stderr)
    end)
  end)

  describe("in a valid project", function()
    local tmp_dir = hh.tmp_dir()
    local project = path.join(tmp_dir, 'project')
    local box     = path.join(tmp_dir, 'test-box')
    local images = {}

    setup(function()
      fs.mkdir(project)
      fs.cp_r(
        hh.fixture_path("full_azkfile.json"),
        path.join(project, azk.manifest)
      )
      fs.cp_r(hh.fixture_path("test-box"), box)
    end)

    it("should provision image-box and image-app before run", function()
      local _, app_data = app.new(project)
      fs.cd(project, function()
        local output = shell.capture_io(function()
          command.run("/bin/bash", "-c", "ls -l")
        end)
        assert.is.match(output.stdout, "total 2.*azkfile%.json")
      end)
    end)
  end)
end)

File: test/commands/exec_spec.js
var h      = require('../spec_helper.js');
var path   = require('path');
var docker = require('../../lib/docker');
var App    = require('../../lib/app');
var cmd    = require('../../lib/commands/exec');

var azk = h.azk;
var Q   = azk.Q;
var _   = azk._;

describe("Azk exec command", function() {
  this.timeout(0);
  var app = null;
  var events, out, outputs = {};

  var mocks = h.mock_outputs(beforeEach, outputs, function() {
    events = [];
    out = h.capture_evs(events);
  });

  before(function() {
    return h.mock_app().then(function(dir) {
      app = new App(dir);
      return app.provision({ force: true }, new h.MemoryStream());
    });
  });

  it("should execute a interactive command", function() {
    var stdin  = new h.MemoryStream();
    var out    = {
      log: function(_, type) {
        if (type == "wait")
          stdin.write("uname\nexit\n");
      }
    }

    var result = cmd.run(app, out, ["/bin/bash"], {
      interactive: true,
      stdin: stdin, stdout: mocks.stdout, stderr: mocks.stderr,
    });

    return result.then(function(code) {
      h.expect(code).to.equal(0);
      h.expect(outputs.stdout).to.match(/Linux/);
    });
  });

  it("should mount app directory", function() {
    var args   = ["/bin/bash", "-c", "cat /azk/app/azkfile.json"];
    var result = cmd.run(app, out, args, {
      stdout: mocks.stdout, stderr: mocks.stderr
    });

    return result.then(function(code) {
      h.expect(code).to.equal(0);
      h.expect(outputs.stdout).to.match(/ENVS_ENV_VAR/);
    });
  });

  it("should return result code", function() {
    var args   = ["/bin/bash", "-c", "exit 127"];
    var result = cmd.run(app, out, args, {});

    return result.then(function(code) {
      h.expect(code).to.equal(127);
    });
  });
});
